Add merge_type parameter to k8s module

willthames opened the pull request:
##### SUMMARY

Allows patching of custom Kubernetes resources that
don't support strategic merge patching

Check that openshift module supports content_type param
(requires version 0.6.1 or newer)

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
k8s

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (devel 9539284010) last updated 2018/07/03 08:58:32 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/opensource/ansible/lib/ansible
  executable location = /Users/will/src/opensource/ansible/bin/ansible
  python version = 2.7.15 (default, May  1 2018, 16:44:08) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)]
```

mattclay commented:
The sanity test failures are valid. The integration test failures are unrelated and can be ignored.

fabianvf commented on lib/ansible/module_utils/k8s/common.py:
```diff
@@ -62,6 +63,9 @@
         'type': 'bool',
         'default': False,
     },
+    'merge_type': {
```
I think this will need to move out of the common arg spec, since it only applies to the k8s module (not scale or probably any future modules). You could implement the argspec property in raw.py, and just add the `merge_type` parameter there

fabianvf replied:
alternatively, you can just pop the key off in the `scale.py` implementation of argspec, it seems to be doing that already.

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -175,7 +175,13 @@ def perform_action(self, resource, definition):
+                    params = dict(name=name, namespace=namespace)
+                    if self.params['merge_type']:
+                        from distutils.version import LooseVersion
+                        if LooseVersion(self.openshift_version) < LooseVersion("0.6.1"):
```
also change this to 0.6.2, just came to my attention that I made an error in the release process and the 0.6.1 release didn't update the openshift.__version__. 0.6.2 is identical to 0.6.1 except with the version properly set. Sorry for the inconvenience

fabianvf commented:
Needs rebase and the merge_type parameter needs to not apply to k8s_scale, but other than that it looks good to me.

willthames committed:
Add merge_type parameter to k8s module

Allows patching of custom Kubernetes resources that
don't support strategic merge patching

Check that openshift module supports content_type param
(requires version newer than 0.6.0)

willthames commented:
@fabianvf @maxamillion ready for rereview

fabianvf commented:
shipit

fabianvf commented:
thanks @willthames
